src/differentiable_functions.jl

```diff
@@ -1,91 +1,159 @@
+                                              reshape_g(g!, initial_x),
+                                              reshape_fg(fg!, initial_x))
+end
+
+function DifferentiableMultivariateFunction(f!, g!)
+    function fg!(x::AbstractVector, fx::AbstractVector, gx::AbstractMatrix)
         f!(x, fx)
         g!(x, gx)
     end
     return DifferentiableMultivariateFunction(f!, g!, fg!)
 end
 
-function DifferentiableMultivariateFunction(f!::Function; dtype::Symbol=:central)
-    function fg!(x::Vector, fx::Vector, gx::Array)
+function DifferentiableMultivariateFunction(f!; dtype::Symbol=:central)
+    function fg!(x::AbstractVector, fx::AbstractVector, gx::AbstractMatrix)
         f!(x, fx)
-        function f(x::Vector)
+        function f(x::AbstractVector)
             fx = similar(x)
             f!(x, fx)
             return fx
         end
         finite_difference_jacobian!(f, x, fx, gx, dtype)
     end
-    function g!(x::Vector, gx::Array)
+    function g!(x::AbstractVector, gx::AbstractMatrix)
         fx = similar(x)
         fg!(x, fx, gx)
     end
     return DifferentiableMultivariateFunction(f!, g!, fg!)
 end
 
+function DifferentiableMultivariateFunction(f!, initial_x::AbstractArray;
+                                            dtype::Symbol=:central)
+    return DifferentiableMultivariateFunction(reshape_f(f!, initial_x); dtype=dtype)
+end
+
 # Helper for the case where only f! and fg! are available
-function only_f!_and_fg!(f!::Function, fg!::Function)
-    function g!(x::Vector, gx::Array)
+function only_f!_and_fg!(f!, fg!)
+    function g!(x::AbstractVector, gx::AbstractMatrix)
         fx = similar(x)
         fg!(x, fx, gx)
     end
     return DifferentiableMultivariateFunction(f!, g!, fg!)
 end
 
+function only_f!_and_fg!(f!, fg!, initial_x::AbstractArray)
+    return only_f!_and_fg!(reshape_f(f!, initial_x),
+                           reshape_fg(fg!, initial_x))
+end
+
 # Helper for the case where only fg! is available
-function only_fg!(fg!::Function)
-    function f!(x::Vector, fx::Vector)
+function only_fg!(fg!)
+    function f!(x::AbstractVector, fx::AbstractVector)
         gx = Array{eltype(x)}(length(x), length(x))
         fg!(x, fx, gx)
     end
-    function g!(x::Vector, gx::Array)
+    function g!(x::AbstractVector, gx::AbstractMatrix)
         fx = similar(x)
         fg!(x, fx, gx)
     end
     return DifferentiableMultivariateFunction(f!, g!, fg!)
 end
 
+function only_fg!(fg!, initial_x::AbstractArray)
+    return only_fg!(reshape_fg(fg!, initial_x))
+end
+
 # For sparse Jacobians
-struct DifferentiableSparseMultivariateFunction <: AbstractDifferentiableMultivariateFunction
-    f!::Function
-    g!::Function
-    fg!::Function
+struct DifferentiableSparseMultivariateFunction{F1,F2,F3} <: AbstractDifferentiableMultivariateFunction
+    f!::F1
+    g!::F2
+    fg!::F3
+
+    DifferentiableSparseMultivariateFunction{F1,F2,F3}(f!, g!, fg!) where {F1,F2,F3} =
```

Comment:
and here..? I know you probably added them for a reason, I'm just not sure why :)

Reply:
The problem was that outer constructors such as `DifferentiableSparseMultivariateFunction(f!, g!, initial_x::AbstractArray)` were never called since these calls were already catched by the default constructors, and apparently it was solved by adding (slightly different) default constructors. However, the correct way is to just specify `DifferentiableSparseMultivariateFunction(f!::F1, g!::F2, initial_x::AbstractArray) where {F1,F2}`, I guess. I'll fix it.